core/src/main/java/io/grpc/ServerImpl.java

```diff
@@ -153,12 +116,11 @@ public synchronized ServerImpl start() {
    * Initiates an orderly shutdown in which preexisting calls continue but new calls are rejected.
    */
   public synchronized ServerImpl shutdown() {
-    shutdown = true;
-    // transports collection can be modified during stopAsync(), even if we hold the lock, due to
-    // reentrancy.
-    for (Service transport : transports.toArray(new Service[transports.size()])) {
-      transport.stopAsync();
+    if (shutdown) {
+      return this;
     }
+    transportServer.shutdown();
+    shutdown = true;
```

Comment:
I'm not sure it matters one way or the other. Since all uses of shutdown have to hold the lock, it could only matter due to reentrancy from the transportServer.shutdown() or due to an exception.

This shutdown only applies to the ServerImpl, not the underlying transports or the transport.Server. We are forced to delay actually performing the rest of the shutdown until ServerListener.serverShutdown() to avoid deadlocks and broken state.